### PR TITLE
Fix resolving binding references to additionalPorts

### PR DESCRIPTION
**Pull request by vhvb1989**
compile all ingress and use it for resolving env

### CODE REVIEW

**Comment by davidfowl** [comment]
Remaining issues:

> The other issue is that when a binding has been mapped to http ingress, the port should be 80 and https 443. The target port should be whatever was specified or inferred by azd.

When resolving any endpoint mapped to the primary http ingress, we need to remove the target port from the URL. This can be seen when trying to publish the [qdrant sample](https://github.com/dotnet/aspire/tree/main/playground/Qdrant):

image

This is incorrect and should have the port removed. 

This is a minimal example:

```C#
var builder = DistributedApplication.CreateBuilder(args);

var p = builder.AddProject<Projects.WebApplication1>("api", launchProfileName: null)
       .WithHttpEndpoint(targetPort: 5000, name: "http")
       .WithExternalHttpEndpoints();

p.WithEnvironment(c =>
{
    c.EnvironmentVariables["http_ep"] = p.GetEndpoint("http");
});

builder.Build().Run();
```

manifest

```JSON
{
  "resources": {
    "api": {
      "type": "project.v0",
      "path": "../WebApplication1/WebApplication1.csproj",
      "env": {
        "OTEL_DOTNET_EXPERIMENTAL_OTLP_EMIT_EXCEPTION_LOG_ATTRIBUTES": "true",
        "OTEL_DOTNET_EXPERIMENTAL_OTLP_EMIT_EVENT_LOG_ATTRIBUTES": "true",
        "OTEL_DOTNET_EXPERIMENTAL_OTLP_RETRY": "in_memory",
        "ASPNETCORE_FORWARDEDHEADERS_ENABLED": "true",
        "http_ep": "{api.bindings.http.url}"
      },
      "bindings": {
        "http": {
          "scheme": "http",
          "protocol": "tcp",
          "transport": "http",
          "targetPort": 5000,
          "external": true
        }
      }
    }
  }
}
```

Today this will produce:

```yaml
template:
    containers:
    - image: {{ .Image }}
      name: api
      env:
      - name: AZURE_CLIENT_ID
        value: {{ .Env.MANAGED_IDENTITY_CLIENT_ID }}
      - name: ASPNETCORE_FORWARDEDHEADERS_ENABLED
        value: "true"
      - name: OTEL_DOTNET_EXPERIMENTAL_OTLP_EMIT_EVENT_LOG_ATTRIBUTES
        value: "true"
      - name: OTEL_DOTNET_EXPERIMENTAL_OTLP_EMIT_EXCEPTION_LOG_ATTRIBUTES
        value: "true"
      - name: OTEL_DOTNET_EXPERIMENTAL_OTLP_RETRY
        value: in_memory
      - name: http_ep
        value: http://api.{{ .Env.AZURE_CONTAINER_APPS_ENVIRONMENT_DEFAULT_DOMAIN }}:5000
```

The target port shouldn't be used here because the exposed port is 80 and 443.

**Comment by davidfowl** [comment]
Other issues:

Resolving targetPort and port seem to be incorrect for the default http/https endpoints that default to a null targetPort (which means azd infers the port). Instead we get 0.

```C#
var builder = DistributedApplication.CreateBuilder(args);

var p = builder.AddProject<Projects.WebApplication1>("api", launchProfileName: null)
       .WithHttpEndpoint(name: "http")
       .WithExternalHttpEndpoints();

p.WithEnvironment(c =>
{
    var ep = p.GetEndpoint("http");
    c.EnvironmentVariables["http_host"] = ep.Property(EndpointProperty.Host);
    c.EnvironmentVariables["http_port"] = ep.Property(EndpointProperty.Port);
    c.EnvironmentVariables["http_targetPort"] = ep.Property(EndpointProperty.TargetPort);
    c.EnvironmentVariables["http_scheme"] = ep.Property(EndpointProperty.Scheme);
    c.EnvironmentVariables["http_ep"] = ep;
});

builder.Build().Run();
```

Manifest:

```JSON
{
  "resources": {
    "api": {
      "type": "project.v0",
      "path": "../WebApplication1/WebApplication1.csproj",
      "env": {
        "OTEL_DOTNET_EXPERIMENTAL_OTLP_EMIT_EXCEPTION_LOG_ATTRIBUTES": "true",
        "OTEL_DOTNET_EXPERIMENTAL_OTLP_EMIT_EVENT_LOG_ATTRIBUTES": "true",
        "OTEL_DOTNET_EXPERIMENTAL_OTLP_RETRY": "in_memory",
        "ASPNETCORE_FORWARDEDHEADERS_ENABLED": "true",
        "http_host": "{api.bindings.http.host}",
        "http_port": "{api.bindings.http.port}",
        "http_targetPort": "{api.bindings.http.targetPort}",
        "http_scheme": "{api.bindings.http.scheme}",
        "http_ep": "{api.bindings.http.url}"
      },
      "bindings": {
        "http": {
          "scheme": "http",
          "protocol": "tcp",
          "transport": "http",
          "external": true
        }
      }
    }
  }
}
```

yaml

```yaml
  template:
    containers:
    - image: {{ .Image }}
      name: api
      env:
      - name: AZURE_CLIENT_ID
        value: {{ .Env.MANAGED_IDENTITY_CLIENT_ID }}
      - name: ASPNETCORE_FORWARDEDHEADERS_ENABLED
        value: "true"
      - name: OTEL_DOTNET_EXPERIMENTAL_OTLP_EMIT_EVENT_LOG_ATTRIBUTES
        value: "true"
      - name: OTEL_DOTNET_EXPERIMENTAL_OTLP_EMIT_EXCEPTION_LOG_ATTRIBUTES
        value: "true"
      - name: OTEL_DOTNET_EXPERIMENTAL_OTLP_RETRY
        value: in_memory
      - name: http_ep
        value: http://api.{{ .Env.AZURE_CONTAINER_APPS_ENVIRONMENT_DEFAULT_DOMAIN }}
      - name: http_host
        value: api
      - name: http_port
        value: "0"
      - name: http_scheme
        value: http
      - name: http_targetPort
        value: "0"
```

**Comment by davidfowl** [comment]
azd also doesn't distinguish between port and targetPort in the manifest:

```C#
var builder = DistributedApplication.CreateBuilder(args);

var p = builder.AddProject<Projects.WebApplication1>("api", launchProfileName: null)
       .WithHttpEndpoint(name: "http", port: 80, targetPort: 5000);

p.WithEnvironment(c =>
{
    var ep = p.GetEndpoint("http");
    c.EnvironmentVariables["http_host"] = ep.Property(EndpointProperty.Host);
    c.EnvironmentVariables["http_port"] = ep.Property(EndpointProperty.Port);
    c.EnvironmentVariables["http_targetPort"] = ep.Property(EndpointProperty.TargetPort);
    c.EnvironmentVariables["http_scheme"] = ep.Property(EndpointProperty.Scheme);
    c.EnvironmentVariables["http_ep"] = ep;
});

builder.Build().Run();
```

Manifest:

```JSON
{
  "resources": {
    "api": {
      "type": "project.v0",
      "path": "../WebApplication1/WebApplication1.csproj",
      "env": {
        "OTEL_DOTNET_EXPERIMENTAL_OTLP_EMIT_EXCEPTION_LOG_ATTRIBUTES": "true",
        "OTEL_DOTNET_EXPERIMENTAL_OTLP_EMIT_EVENT_LOG_ATTRIBUTES": "true",
        "OTEL_DOTNET_EXPERIMENTAL_OTLP_RETRY": "in_memory",
        "ASPNETCORE_FORWARDEDHEADERS_ENABLED": "true",
        "http_host": "{api.bindings.http.host}",
        "http_port": "{api.bindings.http.port}",
        "http_targetPort": "{api.bindings.http.targetPort}",
        "http_scheme": "{api.bindings.http.scheme}",
        "http_ep": "{api.bindings.http.url}"
      },
      "bindings": {
        "http": {
          "scheme": "http",
          "protocol": "tcp",
          "transport": "http",
          "port": 80,
          "targetPort": 5000
        }
      }
    }
  }
}
```

yaml

```yaml
  template:
    containers:
    - image: {{ .Image }}
      name: api
      env:
      - name: AZURE_CLIENT_ID
        value: {{ .Env.MANAGED_IDENTITY_CLIENT_ID }}
      - name: ASPNETCORE_FORWARDEDHEADERS_ENABLED
        value: "true"
      - name: OTEL_DOTNET_EXPERIMENTAL_OTLP_EMIT_EVENT_LOG_ATTRIBUTES
        value: "true"
      - name: OTEL_DOTNET_EXPERIMENTAL_OTLP_EMIT_EXCEPTION_LOG_ATTRIBUTES
        value: "true"
      - name: OTEL_DOTNET_EXPERIMENTAL_OTLP_RETRY
        value: in_memory
      - name: http_ep
        value: http://api.internal.{{ .Env.AZURE_CONTAINER_APPS_ENVIRONMENT_DEFAULT_DOMAIN }}:5000
      - name: http_host
        value: api
      - name: http_port
        value: "5000"
      - name: http_scheme
        value: http
      - name: http_targetPort
        value: "5000"
```

notice http_port is 5000 and http_targetPort is 5000

**Comment by davidfowl** [comment]
All of the scenarios aren't fully working.

File: cli/azd/pkg/containerapps/container_app.go [hunk 1]
```diff
@@ -108,92 +101,39 @@ func (cas *containerAppService) GetIngressConfiguration(
 	}, nil
 }
 
-// apiVersionKey is the key that can be set in the root of a deployment yaml to control the API version used when creating
```
Comment (weikanglim) [inline cli/azd/pkg/containerapps/container_app.go:111]: I think this may have came in from a bad merge?

**Review by davidfowl — APPROVED** [review]
Works! There are a few small things that can be follow up:
1. Error messages could use some work
2. I get `{{ targetPortOrDefault 8080 }}` vs `{{ targetPortOrDefault 0 }}` when resolving the target port for projects. This should be consistent.

**Pull request merged**